### PR TITLE
ADD valgrind to gitactions

### PR DESCRIPTION
**Pull request by fgalan**
- [x] FT_FROM_IX and FT_TO_IX support
- [x] Check if ADD CB_NO_CACHE=ON works also in this case, to add these new cases to the matrix

### CODE REVIEW

**Comment by fgalan** [comment]
First pass of valgrind failed here after 76 minutes:

```
Test 187/1190: 0322_multitenancy/tenant_in_notification .............................................................................................  FAILURE! (no valgrind output for functional test tenant_in_notification.test)
```

Relaunching to confirm it.

**Comment by fgalan** [comment]
Second pass of valgrind failed in the same test after 75 minutes:

```
Test 187/1190: 0322_multitenancy/tenant_in_notification .............................................................................................  FAILURE! (no valgrind output for functional test tenant_in_notification.test)
```

To be investigated.

**Comment by fgalan** [comment]
valgrind optimization done in PR https://github.com/telefonicaid/fiware-orion/pull/3920 has been incorporated to this PR in the master sync at commit 3bc9177 

However, maybe something has broken in the build process. The valgrind pass doesn't start and GitActions reports:

```
===================================== MAKE =============================================
Builder: fix makefile
Builder: build valgrind
make: *** No rule to make target 'build_valgrind'.  Stop.
Builder: make failed
Error: Process completed with exit code 1.
```

**Comment by fgalan** [comment]
> However, maybe something has broken in the build process. The valgrind pass doesn't start and GitActions reports:

The fail could be due to CI image is not sync with master so the modifications done in this PR for ci/rpm8 has lost. Surprisingly, new commits in this PR are not triggering ci/rpm8, only the initial commit. Maybe I need to "touch" some file in ci/rpm8 to trigger the build.

**Comment by fgalan** [comment]
> Maybe I need to "touch" some file in ci/rpm8 to trigger the build.

Touch done (along with a small typo fix in comment :) in commit 9c52c46. Indeed, the CI build has been triggered.

**Comment by fgalan** [comment]
After the last changes, execution stops in test 187, so same as before).

```
Test 187/1189: 0322_multitenancy/tenant_in_notification .............................................................................................  FAILURE! (no valgrind output for functional test tenant_in_notification.test)
```

**Comment by fgalan** [comment]
The problem that blocked tests at 187 has been solved by commit 8ff1f42. After that, all the test runs and the process ends successfully after 246 mins (~4 hours).

I'll try now to implement FT_FROM_IX and FT_TO_IX in order to split the work into pieces. Four passes of ~1 hour could be acceptable.

**Comment by fgalan** [comment]
Part of the original content of this PR has been moved to PR https://github.com/telefonicaid/fiware-orion/pull/3931 in order to have an unified CI image with valgrind capabilities. Thus, the part pending in this PR is only the one related with the valgrind workflow in gitactions.

**Comment by fgalan** [comment]
First pass at 7359b28 reports errors in valgring-1 but they are due to https://github.com/telefonicaid/fiware-orion/issues/3924, so we don't worry about them.

It also reports errors in valgrind-4, in particular this one:

```
1 test cases show valgrind errors:
  932: empty_maps_in_metrics shows 4 valgrind errors
```

Doing a new pass in order to check if that is a glitch...

File: .github/workflows/valgrind.yml [hunk 1]
```diff
@@ -0,0 +1,152 @@
+name: Valgrind Tests
```
Comment (fgalan) [inline .github/workflows/valgrind.yml:1]: Optimize this .yml in a similar way than PR https://github.com/telefonicaid/fiware-orion/pull/3932

**Comment by fgalan** [comment]
In a second pass on 7359b28 we see again:

```
 1 functional tests failed (not a leak, just a func-test failure):
   1097: notification_problem_with_custom_header.test (exit code 1)
```

So probably it's not a glitch.

**Comment by fgalan** [comment]
It gets confirmed:

imagen

imagen

**Comment by fgalan** [comment]
From cases/2846_empty_maps_in_metrics/empty_maps_in_metrics.valgrind.out logs:

```
==2870== Conditional jump or move depends on uninitialised value(s)
==2870==    at 0x650DB3: lmTraceGet(char*, int) (in /usr/local/bin/contextBroker)
==2870==    by 0x4AF05B: logTraceTreat(ConnectionInfo*, int, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, ParseData*) (in /usr/local/bin/contextBroker)
==2870==    by 0x527CF7: restService(ConnectionInfo*, RestService*) (in /usr/local/bin/contextBroker)
==2870==    by 0x5291FE: orion::requestServe[abi:cxx11](ConnectionInfo*) (in /usr/local/bin/contextBroker)
==2870==    by 0x521B55: connectionTreat(void*, MHD_Connection*, char const*, char const*, char const*, char const*, unsigned long*, void**) (in /usr/local/bin/contextBroker)
==2870==    by 0x66130E: call_connection_handler (connection.c:2220)
==2870==    by 0x662E57: MHD_connection_handle_idle (connection.c:3513)
==2870==    by 0x66481D: call_handlers (daemon.c:1215)
==2870==    by 0x667762: thread_main_handle_connection (daemon.c:2144)
==2870==    by 0x5E2A149: start_thread (in /usr/lib64/libpthread-2.28.so)
==2870==    by 0x8323DC2: clone (in /usr/lib64/libc-2.28.so)
==2870==  Uninitialised value was created by a stack allocation
==2870==    at 0x650D10: lmTraceGet(char*, int) (in /usr/local/bin/contextBroker)
==2870==
==2870== Conditional jump or move depends on uninitialised value(s)
==2870==    at 0x650DEE: lmTraceGet(char*, int) (in /usr/local/bin/contextBroker)
==2870==    by 0x4AF05B: logTraceTreat(ConnectionInfo*, int, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, ParseData*) (in /usr/local/bin/contextBroker)
==2870==    by 0x527CF7: restService(ConnectionInfo*, RestService*) (in /usr/local/bin/contextBroker)
==2870==    by 0x5291FE: orion::requestServe[abi:cxx11](ConnectionInfo*) (in /usr/local/bin/contextBroker)
==2870==    by 0x521B55: connectionTreat(void*, MHD_Connection*, char const*, char const*, char const*, char const*, unsigned long*, void**) (in /usr/local/bin/contextBroker)
==2870==    by 0x66130E: call_connection_handler (connection.c:2220)
==2870==    by 0x662E57: MHD_connection_handle_idle (connection.c:3513)
==2870==    by 0x66481D: call_handlers (daemon.c:1215)
==2870==    by 0x667762: thread_main_handle_connection (daemon.c:2144)
==2870==    by 0x5E2A149: start_thread (in /usr/lib64/libpthread-2.28.so)
==2870==    by 0x8323DC2: clone (in /usr/lib64/libc-2.28.so)
==2870==  Uninitialised value was created by a stack allocation
==2870==    at 0x650D10: lmTraceGet(char*, int) (in /usr/local/bin/contextBroker)

==2870== Conditional jump or move depends on uninitialised value(s)
==2870==    at 0x650DEE: lmTraceGet(char*, int) (in /usr/local/bin/contextBroker)
==2870==    by 0x4AF05B: logTraceTreat(ConnectionInfo*, int, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, ParseData*) (in /usr/local/bin/contextBroker)
==2870==    by 0x527CF7: restService(ConnectionInfo*, RestService*) (in /usr/local/bin/contextBroker)
==2870==    by 0x5291FE: orion::requestServe[abi:cxx11](ConnectionInfo*) (in /usr/local/bin/contextBroker)
==2870==    by 0x521B55: connectionTreat(void*, MHD_Connection*, char const*, char const*, char const*, char const*, unsigned long*, void**) (in /usr/local/bin/contextBroker)
==2870==    by 0x66130E: call_connection_handler (connection.c:2220)
==2870==    by 0x662E57: MHD_connection_handle_idle (connection.c:3513)
==2870==    by 0x66481D: call_handlers (daemon.c:1215)
==2870==    by 0x667762: thread_main_handle_connection (daemon.c:2144)
==2870==    by 0x5E2A149: start_thread (in /usr/lib64/libpthread-2.28.so)
==2870==    by 0x8323DC2: clone (in /usr/lib64/libc-2.28.so)
==2870==  Uninitialised value was created by a stack allocation
==2870==    at 0x650D10: lmTraceGet(char*, int) (in /usr/local/bin/contextBroker)
==2870==
==2870==
==2870== 2 errors in context 2 of 5:
==2870== Conditional jump or move depends on uninitialised value(s)
==2870==    at 0x650DB3: lmTraceGet(char*, int) (in /usr/local/bin/contextBroker)
==2870==    by 0x4AF05B: logTraceTreat(ConnectionInfo*, int, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, ParseData*) (in /usr/local/bin/contextBroker)
==2870==    by 0x527CF7: restService(ConnectionInfo*, RestService*) (in /usr/local/bin/contextBroker)
==2870==    by 0x5291FE: orion::requestServe[abi:cxx11](ConnectionInfo*) (in /usr/local/bin/contextBroker)
==2870==    by 0x521B55: connectionTreat(void*, MHD_Connection*, char const*, char const*, char const*, char const*, unsigned long*, void**) (in /usr/local/bin/contextBroker)
==2870==    by 0x66130E: call_connection_handler (connection.c:2220)
==2870==    by 0x662E57: MHD_connection_handle_idle (connection.c:3513)
==2870==    by 0x66481D: call_handlers (daemon.c:1215)
==2870==    by 0x667762: thread_main_handle_connection (daemon.c:2144)
==2870==    by 0x5E2A149: start_thread (in /usr/lib64/libpthread-2.28.so)
==2870==    by 0x8323DC2: clone (in /usr/lib64/libc-2.28.so)
==2870==  Uninitialised value was created by a stack allocation
==2870==    at 0x650D10: lmTraceGet(char*, int) (in /usr/local/bin/contextBroker)
```

File: .github/workflows/valgrind.yml (hunk 1 above)
Comment (fgalan) [reply .github/workflows/valgrind.yml:1]: Fixed in ae779cc5387faeb36c7e6a718def3b6edab24277

**Comment by fgalan** [comment]
Hacking the ci/rpm8/makefile this way to force build in debug mode (attempting to get the exact code line in which is the problem):

```diff
--- a/ci/rpm8/makefile
+++ b/ci/rpm8/makefile
@@ -74,15 +74,15 @@ unit:
 build_functional: prepare
        @echo '------------------------------------- make build_functional ----------------------------'
 
-       rm -Rf BUILD_FUNCTIONAL || true && mkdir -p BUILD_FUNCTIONAL || true
+       rm -Rf BUILD_DEBUG || true && mkdir -p BUILD_DEBUG || true
 
-       cd BUILD_FUNCTIONAL && cmake .. -DCMAKE_BUILD_TYPE=RELEASE -DBUILD_ARCH=$(BUILD_ARCH) -DCMAKE_INSTALL_PREFIX=$(INSTALL_DIR)
-       cd BUILD_FUNCTIONAL && make -j$(CPU_COUNT)
+       cd BUILD_DEBUG && cmake .. -DCMAKE_BUILD_TYPE=DEBUG -DBUILD_ARCH=$(BUILD_ARCH) -DCMAKE_INSTALL_PREFIX=$(INSTALL_DIR)
+       cd BUILD_DEBUG && make -j$(CPU_COUNT)
 
 install_functional: install
        @echo '------------------------------------- make install_functional --------------------------'
 
-       cd BUILD_FUNCTIONAL && make install DESTDIR=$(DESTDIR)
+       cd BUILD_DEBUG && make install DESTDIR=$(DESTDIR)
 
 functional:
        @echo '------------------------------------- make functional ----------------------------------'
```

There is no error reported:

```
[root@debian11 opt]# FT_FROM_IX=932 FT_TO_IX=932 build -miqts valgrind
...
Install the project...
-- Install configuration: "DEBUG"
-- Installing: //usr/local/bin/contextBroker
-- Set runtime path of "//usr/local/bin/contextBroker" to ""
make[1]: Leaving directory '/opt/fiware-orion/BUILD_DEBUG'
Builder: revert makefile
===================================== VALGRIND TESTS =================================
Builder: fix makefile
Thu Sep  9 11:23:18 UTC 2021
Run tests 932 to 932
Test 932/1201: 2846_empty_maps_in_metrics/empty_maps_in_metrics ..................................................................................... OK (8.68 seconds)
Total test time: 9.83 seconds (0 minutes)
Great, all valgrind tests ran without any memory leakage
Builder: revert makefile
```

This makes the problem more difficult to solve. The only thing we now is that some uninitialised variable in the stack defined lmTraceGet() is causing problems, but that function is ~70 and it's not evident which one is. Any clue @kzangeli ?

**Comment by fgalan** [comment]
With regards to CB_NO_CACHE=ON support:

-------------

If I run

```
./valgrindTestSuite.sh 
```

I can check in a separate terminal that contextBroker is started **without** -noCache CLI flag

```
$ ps ax | grep context
 827795 ?        Rsl    0:01 /usr/bin/valgrind.bin -v --leak-check=full --track-origins=yes --trace-children=yes contextBroker -harakiri -port 9999 -pidpath /tmp/orion_9999.pid -dbhost localhost:27017 -db ftest -dbPoolSize 10 -t 0-255 -logLevel DEBUG -notificationMode threadpool:200:5
```

-------------

If I run

```
CB_NO_CACHE=ON ./valgrindTestSuite.sh
```

I can check in a separate terminal that contextBroker is started **with** -noCache CLI flag

```
$ ps ax | grep context
 828262 ?        Ssl    0:01 /usr/bin/valgrind.bin -v --leak-check=full --track-origins=yes --trace-children=yes contextBroker -harakiri -port 9999 -pidpath /tmp/orion_9999.pid -dbhost localhost:27017 -db ftest -dbPoolSize 10 -t 0-255 -logLevel DEBUG -noCache -notificationMode threadpool:200:5

```

----------

Conclusion: CB_NO_CACHE=ON works as expected

**Comment by fgalan** [comment]
At commit 7e10892 (which corresponds to a release candidate for 3.4.0) only the already known empty_maps_in_metrics.test have fails. Thus, the release candidate is suitable for releaseing from a valgrind point of view.

**Comment by fgalan** [comment]
At commit c1c7d6c (which corresponds to a release candidate for 3.5.0) only the already known empty_maps_in_metrics.test have fails. Thus, the release candidate is suitable for releaseing from a valgrind point of view.

**Comment by fgalan** [comment]
At commit [52367dc](https://github.com/telefonicaid/fiware-orion/pull/3880/commits/52367dc3f34f7ddbcb1c0597dceca03a9f2b6d59) (which corresponds to a release candidate for 3.6.0) only the already known empty_maps_in_metrics.test have fails. Thus, the release candidate is suitable for releaseing from a valgrind point of view.

**Comment by fgalan** [comment]
At commit [eb09e44](https://github.com/telefonicaid/fiware-orion/pull/3880/commits/eb09e447df7b663855e277548c025314d6a19585) (which corresponds to a release candidate for 3.7.0) only the already known empty_maps_in_metrics.test have fails. Thus, the release candidate is suitable for releasing from a valgrind point of view.

File: .github/workflows/valgrind.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+on:
+  push:
+  pull_request:
+    branches:
+      - master
+
+env:
+   TEST_IMAGE_NAME: fiware/orion-ci:rpm8
```
Comment (fgalan) [inline .github/workflows/valgrind.yml:10]:
```suggestion
   TEST_IMAGE_NAME: fiware/orion-ci:deb
```

**Comment by fgalan** [comment]
I think I have narrowed down the problem that cause the weird valgrind error to this fragment of code in the `lmTraceGet()` function.

```
  for (i = 1; i < j; i++)
  {
    int   prev   = levels[i - 1];
    int   diss   = levels[i];
    int   next   = levels[i + 1];
    bool  before = (diss == (prev + 1));
    bool  after  = (diss == (next - 1));
    char  str[12];

    if (i == 255)
      after = false;

    if (before && after)
    {
    }
    else if (before && !after)
    {
      snprintf(str, sizeof(str), "-%d", diss);
      strncat(levelString, str, levelStringSize - strlen(levelString) - 1);
    }
    else if (!before && after)
    {
      snprintf(str, sizeof(str), ", %d", diss);
      strncat(levelString, str, levelStringSize - strlen(levelString) - 1);
    }
    else if (!before && !after)
    {
      snprintf(str, sizeof(str), ", %d", diss);
      strncat(levelString, str, levelStringSize - strlen(levelString) - 1);
    }
  }
```

I have empirically checked that if that fragment is disabled, valgrind stops complaining (although the test breaks, of course).

**Comment by fgalan** [comment]
This modification in the test makes solves valgrind in empty_maps_in_metrics.test

```diff
diff --git a/test/functionalTest/cases/2846_empty_maps_in_metrics/empty_maps_in_metrics.test b/test/functionalTest/cases/2846_empty_maps_in_metrics/empty_maps_in_metrics.test
index 4dabf7cad..d40fd5427 100644
--- a/test/functionalTest/cases/2846_empty_maps_in_metrics/empty_maps_in_metrics.test
+++ b/test/functionalTest/cases/2846_empty_maps_in_metrics/empty_maps_in_metrics.test
@@ -25,7 +25,7 @@ Metrics without emptied reported items that stay
 
 --SHELL-INIT--
 dbInit CB
-brokerStart CB
+brokerStart CB 0-200
 
 --SHELL--
 
@@ -74,7 +74,7 @@ Fiware-Correlator: REGEX([0-9a-f\-]{36})
 Date: REGEX(.*)
 
 {
-    "tracelevels": "0-255"
+    "tracelevels": "0-200"
 }
 
 
@@ -87,7 +87,7 @@ Fiware-Correlator: REGEX([0-9a-f\-]{36})
 Date: REGEX(.*)
 
 {
-    "tracelevels": "0-255"
+    "tracelevels": "0-200"
 }
```

So I guess the problems is some kind of "edge effect" with the 255 tracelevel.

**Review by mapedraza — APPROVED** [review]
LGTM

**Pull request merged**